src/dbnode/persist/fs/retriever.go

```diff
@@ -431,48 +448,58 @@ func (r *blockRetriever) fetchBatch(
 		req.indexEntry = entry
 	}
 
-	sort.Sort(retrieveRequestByOffsetAsc(reqs))
+	sort.Sort(retrieveRequestByIndexEntryOffsetAsc(reqs))
 	tagDecoderPool := r.fsOpts.TagDecoderPool()
 
 	blockCachingEnabled := r.opts.CacheBlocksOnRetrieve() && r.nsCacheBlocksOnRetrieve
 
 	// Seek and execute all requests
 	for _, req := range reqs {
-		var (
-			data checked.Bytes
-			err  error
-		)
+		// Should always be a data request by this point.
+		if req.streamReqType != streamDataReq {
+			req.err = fmt.Errorf("wrong stream req type: expect=%d, actual=%d",
+				streamDataReq, req.streamReqType)
+			continue
+		}
 
-		// Only try to seek the ID if it exists and there haven't been any errors so
-		// far, otherwise we'll get a checksum mismatch error because the default
-		// offset value for indexEntry is zero.
-		if req.foundAndHasNoError() {
-			data, err = seeker.SeekByIndexEntry(req.indexEntry, seekerResources)
-			if err != nil && err != errSeekIDNotFound {
-				req.onError(err)
-				continue
-			}
+		if req.err != nil {
+			// Skip requests with error, will already get appropriate callback.
+			continue
+		}
+
+		if req.notFound {
+			// Only try to seek the ID if it exists and there haven't been any errors so
+			// far, otherwise we'll get a checksum mismatch error because the default
+			// offset value for indexEntry is zero.
+			req.success = true
+			req.onCallerOrRetrieverDone()
+			continue
+		}
+
+		data, err := seeker.SeekByIndexEntry(req.indexEntry, seekerResources)
+		if err != nil {
+			// If not found error is returned here, that's still an error since
+			// it's expected to be found if it was found in the index file.
+			req.err = err
+			continue
 		}
 
 		var (
 			seg, onRetrieveSeg ts.Segment
 			checksum           = req.indexEntry.DataChecksum
 		)
-		if data != nil {
-			seg = ts.NewSegment(data, nil, checksum, ts.FinalizeHead)
-		}
+		seg = ts.NewSegment(data, nil, checksum, ts.FinalizeHead)
```

Comment:
Yeah, so before the flow was honestly a little fragile, as `data` was assigned only in certain conditions (e.g. it could be nil if the request was not "found"), so we had to do the nil check. After the refactor, `data` comes from `seeker.SeekByIndexEntry`; the only time it's nil is if the seek itself errors, but we just skip the entry in that case